algebra/ring.lean

```diff
@@ -57,6 +57,151 @@ section comm_ring
   ⟨dvd_of_neg_dvd, neg_dvd_of_dvd⟩
 end comm_ring
 
+def is_unit {α : Type u} [comm_ring α] (x : α) := ∃ y, x * y = 1
+def nonunits (α : Type u) [comm_ring α] : set α := { x | ¬∃ y, x * y = 1 }
+
+class is_hom {α : Type u} {β : Type v} [comm_ring α] [comm_ring β] (f : α → β) : Prop :=
+(map_add : ∀ {x y}, f (x + y) = f x + f y)
+(map_mul : ∀ {x y}, f (x * y) = f x * f y)
+(map_one : f 1 = 1)
+
+namespace is_hom
+
+variables {α : Type u} {β : Type v} [comm_ring α] [comm_ring β]
+variables (f : α → β) [is_hom f] {x y : α}
+
+attribute [simp] map_add
+attribute [simp] map_mul
+attribute [simp] map_one
+
+@[simp] lemma map_zero : f 0 = 0 :=
+calc f 0 = f (0 + 0) - f 0 : by rw [map_add f]; simp
+     ... = 0 : by simp
+
+@[simp] lemma map_neg : f (-x) = -f x :=
+calc f (-x) = f (-x + x) - f x : by rw [map_add f]; simp
+        ... = -f x : by simp [map_zero f]
+
+@[simp] lemma map_sub : f (x - y) = f x - f y :=
+by simp [map_add f, map_neg f]
+
+end is_hom
+
+class is_ideal (α : Type u) [comm_ring α] (S : set α) : Prop :=
+(zero_mem : (0 : α) ∈ S)
+(add_mem : ∀ {x y}, x ∈ S → y ∈ S → x + y ∈ S)
+(mul_mem : ∀ {x y}, x ∈ S → x * y ∈ S)
+
+namespace is_ideal
+
+variables {α : Type u} [comm_ring α] {S : set α} [is_ideal α S] {x y z : α}
+include S
+
+attribute [simp] zero_mem
+
+lemma neg_mem : x ∈ S → -x ∈ S :=
+λ hx, have h : x * -1 ∈ S, from is_ideal.mul_mem hx, by simpa using h
+
+lemma sub_mem : x ∈ S → y ∈ S → x - y ∈ S :=
+λ hx hy, have h : x + -y ∈ S, from add_mem hx $ neg_mem hy, by simpa using h
+
+lemma mem_of_mem_of_add_mem : x ∈ S → x + y ∈ S → y ∈ S :=
+λ hx hxy, calc
+    y = y + x - x : eq.symm $ add_sub_cancel y x
+  ... = x + y - x : congr_arg (λ z, z - x) $ add_comm y x
+  ... ∈ S : sub_mem hxy hx
+
+lemma mem_of_mem_of_add_mem' : y ∈ S → x + y ∈ S → x ∈ S :=
+λ hy hxy, calc
+    x = x + y - y : eq.symm $ add_sub_cancel x y
+  ... ∈ S : sub_mem hxy hy
+
+lemma mem_of_mem_of_sub_mem : x ∈ S → x - y ∈ S → y ∈ S :=
+λ hx hxy, calc
+    y = x - (x - y) : eq.symm $ sub_sub_cancel x y
+  ... ∈ S : sub_mem hx hxy
+
+lemma mem_of_mem_of_sub_mem' : y ∈ S → x - y ∈ S → x ∈ S :=
+λ hy hxy, calc
+    x = x - y + y : eq.symm $ sub_add_cancel x y
+  ... ∈ S : add_mem hxy hy
+
+lemma mul_mem' : y ∈ S → x * y ∈ S :=
+λ hy, have h : y * x ∈ S, from mul_mem hy, by rwa [mul_comm]
+
+end is_ideal
+
+@[reducible] def zero_ideal (α : Type u) [comm_ring α] : set α := {(0:α)}
+instance zero_ideal.is_ideal (α : Type u) [comm_ring α] : is_ideal α $ zero_ideal α :=
+by refine {..}; intros; simp [set.mem_singleton_iff] at *; simp [*]
+
+@[reducible] def univ_ideal (α : Type u) [comm_ring α] : set α := set.univ
+instance univ_ideal.is_ideal (α : Type u) [comm_ring α] : is_ideal α $ univ_ideal α :=
+by refine {..}; intros; trivial
```

Comment:
I don't see why `univ_ideal` exists. It's the same as `univ`. (You can have an instance saying that `univ` is an ideal without giving it a new name.)